xarray/core/dataset.py

```diff
@@ -1110,16 +1122,90 @@ def relevant_keys(mapping):
                 # dim already exists
                 raise ValueError('Existing dimensions are not valid choices '
                                  'for the dim argument in sel_points')
+            # replace the array dim
+            dim_coord = dim.copy()
+            dim = dim.name
 
         if not utils.is_scalar(dim) and not isinstance(dim, DataArray):
-            dim = as_variable(dim, name='points')
+            name = 'points' if not hasattr(dim, 'name') else dim.name
+            dim_coord = as_variable(dim, name=name)
+            dim = name
+
+        indexers_dict = dict(indexers)
+        non_indexed_dims = set(self.dims) - indexer_dims
+
+        sel_coords = OrderedDict()
+        if dim_coord is not None:
+            sel_coords[dim] = dim_coord
+
+        # Add sliced versions of coordinates that themselves depend on other dimensions
+        # which may also have been indexed
+        for c in self.coords:
+            if c in indexers_dict:
+                selection = take(self.coords[c], indexers_dict[c])
+                sel_coords[c] = (dim, selection)
+
+            elif any(d in indexer_dims for d in self.coords[c].dims):
+                # If the coord variable depends on an indexed dimension, slice it
+                var = self.coords[c]
+                coord_dim = var.dims[0]  # should just be one?
+                selection = take(var, indexers_dict[coord_dim])
+
+                sel_coords[c] = (dim, selection)
+            else:
+                sel_coords[c] = self.coords[c]
+
+        variables = OrderedDict()
+
+        for name in data_vars:
+            var = self.variables[name]
+            # Transpose the var to ensure that the indexed dims come first
+            # These dims will be collapsed in the output.
+            # To avoid edge cases in numpy want to transpose to ensure the indexed dimensions are first
+            # However transpose is not lazy, so want to avoid using it for dask case (??)
```

Comment:
This is a potential concern if the data is *not* loaded in dask but is still pointing to a lazily loaded array of values on disk in a netCDF file -- calling `.transpose` will load the entire array into memory.

That said, I think this change is still worth doing, and we shouldn't add special cases for different data types. It is not unreasonable to encourage users to use dask if they are concerned about memory usage and lazy loading.